First version of the SLD Generator Maven Plugin. SldUtils.java and th…

wtenbosch opened the pull request:
First version of the SLD Generator Maven Plugin. SldUtils.java and their related test files where copied from an old version of AERIUS-II. The reading of JSON files (with databinding) and the Mojo-plugin are newly developed.

wtenbosch committed:
First version of the SLD Generator Maven Plugin. SldUtils.java and their related test files where copied from an old version of AERIUS-II. The reading of JSON files (with databinding) and the Mojo-plugin are newly developed.

Hilbrand commented on README.md:
```diff
@@ -0,0 +1,117 @@
+
+## About
+
+Geoserver uses `.sld` files as stylesheets for the layers. These stylesheets specify which features are visible and with which style. These stylesheets are large `.xml` files, and their syntax is cumbersome.
```
Can you put each sentence on a single line. This would make reviewing and updating lines easier.

SerhatG commented on pom.xml:
```diff
@@ -0,0 +1,135 @@
+
+  <build>
+    <plugins>
+      <plugin>
+        <artifactId>maven-compiler-plugin</artifactId>
+        <version>3.8.1</version>
+        <configuration>
+          <source>${java.version}</source>
+          <target>${java.version}</target>
+          <useIncrementalCompilation>false</useIncrementalCompilation>
+        </configuration>
+      </plugin>
```
This can go if you replace the `java.version` property with:
```xml
    <maven.compiler.source>11</maven.compiler.source>
    <maven.compiler.target>11</maven.compiler.target>
```
The maven-compiler-plugin will use those properties to determine it's internal configuration.

wtenbosch replied:
I didn't know this, thanks!

Hilbrand commented on README.md:
````diff
@@ -0,0 +1,117 @@
+}
+```
+
+* `zoomLevelsFile` refers to the zoom level definition file. The zoom level definition is separate because it can be shared between styles. If no zoom level file is defined, no filtering takes place on zoom level.
+* `rules` array containing the rules
+  * `condition` contains the boolean condition to render this style to a feature
+  * `fillColor` color to fill the feature (hexadecimal)
+  * `strokeColor` color to outline the feature (hexadecimal)
+  * `imageUrl` url to an image to draw at the center of the feature. The url is relative to the final `.sld` file, can be a local file in geoserver, or an external image (specify full url)
+  * `customDrawSld` custom XML used in the draw part for this rule in the `.sld` file (for advanced options which can not be specified using above properties)
+  * `customConditionSld` custom XML used in the condition part for this rule in the `.sld` file (for advanced options which can not be specified using above properties)
````
Maybe put in a table:


| key | Description |
|--|--|
|`condition` | contains the boolean condition to render this style to a feature |
...

Hilbrand commented on pom.xml:
```diff
@@ -0,0 +1,135 @@
+      <name>AERIUS Nexus repository</name>
+      <url>https://nexus.aerius.nl/repository/maven-public/</url>
+    </repository>
+    <repository>
+      <id>OSGEO</id>
+      <name>OSGEO GeoServer Maven Repository</name>
+      <url>https://repo.osgeo.org/repository/release/</url>
+    </repository>
```
Is the OSGEO repository needed?

wtenbosch replied:
No, not needed. I used the pom of geoserver-core as basis, that's why it was here in the first place. Removed.

Hilbrand commented on src/main/java/nl/aerius/sldgenerator/plugin/FileScanner.java:
```diff
@@ -0,0 +1,47 @@
+          .map(Path::toFile);
+
+    } catch (IOException e) {
+      e.printStackTrace();
```
Can the application continue if IOException happens here? Maybe throw the exception, or use logger, but don't use printStackTrace

wtenbosch replied:
Moved the IOException to the Mojo class where it is wrapped in a maven exception. The build should fail if there is an error generating sld's because layers might break unnoticed otherwise.

Hilbrand commented on src/main/java/nl/aerius/sldgenerator/plugin/SldPlugin.java:
```diff
@@ -0,0 +1,68 @@
+    } catch (JacksonException e) {
+      log.error("JSON Parse Error in SLD Generation");
+      log.error(e);
+    } catch (IOException e) {
```
It also throws the IOException, maybe not catch it here?

wtenbosch replied:
Moved the IOException to the Mojo class where it is wrapped in a maven exception. The build should fail if there is an error generating sld's because layers might break unnoticed otherwise.

Hilbrand commented on src/main/java/nl/aerius/sldgenerator/plugin/SldPlugin.java:
```diff
@@ -0,0 +1,68 @@
+      targetFile.write(generatedSld.getBytes(StandardCharsets.UTF_8));
+
+    } catch (JacksonException e) {
+      log.error("JSON Parse Error in SLD Generation");
```
Can you add some context information, maybe `sourceFilePath` and does it make sense to continue or maybe not catch it here?

wtenbosch replied:
Moved the IOException to the Mojo class where it is wrapped in a maven exception. The build should fail if there is an error generating sld's because layers might break unnoticed otherwise.

Added the file path. A detailed error of why it can not be parsed is already in the description of the exception. (e.g. line number and property).

wtenbosch committed:
Process review comments

SerhatG committed:
Add GitHub actions to run Sonar and publish artifact to Sonar

SerhatG commented:
I added a commit to run Sonar and to publish the plugin to Nexus.
Some additional changes might be needed for the coverage part. @Hilbrand knows more about that.

Hilbrand committed:
Added spotless and sonar configurations

- Added AERIUS spotless/header configuration.
- Added sonar profile to run sonar only under a specific profile.
- Fixed build.yml action -> should point to main branch

Hilbrand reviewed (commented):
I've added the spotless/sonar configuration in a pr to this pr: https://github.com/wtenbosch/sld-generator-maven-plugin/pull/1

wtenbosch committed:
Add unit test for FileScanner, move test to correct package

wtenbosch committed:
Merge pull request #1 from Hilbrand/1

Added spotless and sonar configurations

wtenbosch committed:
Merge branch 'first-version' of github.com:wtenbosch/sld-generator-maven-plugin into first-version

wtenbosch committed:
Missing copyright header

SerhatG reviewed (approved):
Let's go for it!
